pkg/workceptor/workceptor.go

```diff
@@ -396,12 +667,23 @@ func (w *Workceptor) GetResults(unitID string) (chan []byte, error) {
 				resultChan <- buf[:n]
 			}
 			if err == io.EOF {
-				if unit.state == WorkStateSucceeded || unit.state == WorkStateFailed {
+				err = stdout.Close()
+				if err != nil {
+					debug.Printf("Error closing stdout\n")
+					return
+				}
+				stat, err := os.Stat(stdoutFilename)
+				var stdoutSize int64
+				if err != nil {
+					stdoutSize = 0
+				} else {
+					stdoutSize = stat.Size()
+				}
+				if IsComplete(unit.status.State) && stdoutSize >= unit.status.StdoutSize {
 					close(resultChan)
-					break
+					debug.Printf("Stdout complete - closing channel\n")
+					return
 				}
-				// Don't spin loop the CPU checking and re-checking for EOF
-				time.Sleep(100 * time.Millisecond)
 				continue
```

Comment:
should we be able to just return here, since we reached EOF?

Reply:
No, because the file is being actively written to by the other process.  Suppose some task has 1500 bytes of stdout that are eventually going to be written.  At the time we do the check it has written 1000 bytes, and ten seconds in the future, will write another 500.  We get an EOF after reading the first 1000 bytes - we did, in fact, get to the end of the file as it currently exists on disk.  But we can't exit yet because we haven't delivered the last 500 bytes.

Terminating the loop first requires that the process is in a completed state (`WorkStateSuccess` or `WorkStateFailed` - this is what `IsComplete` checks).  In this state we know that the value of `unit.status.StdoutSize` is no longer changing, so once our own count of bytes written reaches that size, we know we have reached the true end of stdout and can close resultsChan.